packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts

```diff
@@ -0,0 +1,322 @@
+   * The underlying CFN GraphQL Schema Resource. This
+   * resource is created on `bind` which allows the Schema
+   * to be created outside the CDK (a feature that works well
+   * with the code-first approach).
+   *
+   * @internal
+   */
+  protected _schema!: CfnGraphQLSchema;
+
+  /**
+   * The SchemaMode utilized by the Schema Implementation
+   *
+   * @internal
+   */
+  protected _mode: SchemaMode;
+
+  public constructor() {
+    this.definition = '';
+    this._mode = SchemaMode.FILE;
+  }
+
+  /**
+   * Called when the GraphQL Api is initialized to allow this object to bind
+   * to the stack.
+   */
+  public bind(api: GraphqlApi): CfnGraphQLSchema {
+    if (!this._schema) {
+      this._schema = new CfnGraphQLSchema(api, 'Schema', {
+        apiId: api.apiId,
+        definition: this.definition,
+      });
+    }
+    return this._schema;
+  }
+
+  /**
+   * Escape hatch to add to Schema as desired. Will always result
+   * in a newline.
+   */
+  public addToSchema(_addition: string, _delimiter?: string): void {
+    throw new Error('API cannot append to schema because schema definition mode is not configured as CODE.');
+  }
+
+  /**
+   * Add a query field to the schema's Query. CDK will create an
+   * Object Type called 'Query'. For example,
+   *
+   * type Query {
+   *   fieldName: Field.returnType
+   * }
+   */
+  public addQuery(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add query. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+   * Add a mutation field to the schema's Mutation. CDK will create an
+   * Object Type called 'Mutation'. For example,
+   *
+   * type Mutation {
+   *   fieldName: Field.returnType
+   * }
+   */
+  public addMutation(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add mutation. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+   * Add a subscription field to the schema's Subscription. CDK will create an
+   * Object Type called 'Subscription'. For example,
+   *
+   * type Subscription {
+   *   fieldName: Field.returnType
+   * }
+   */
+  public addSubscription(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add subscription. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+   * Add type to the schema
+   */
+  public addType(_type: IIntermediateType): IIntermediateType {
+    throw new Error('API cannot add type because schema definition mode is not configured as CODE.');
+  }
+}
+
+/**
+ * Schema Implementation for a Schema generated through an Asset
+ */
+export class AssetSchemaImpl extends SchemaImpl {
+  /**
+   * The definition for this schema
+   */
+  public readonly definition: string;
+
+  public constructor(filePath: string) {
+    super();
+    this.definition = readFileSync(filePath).toString('utf-8');
+  }
+}
```

Comment:
Sorry, maybe I'm missing something.

Why does the `AssetSchema` read the file directly from disk? Doesn't this mean the Asset is not actually needed? (Assets are actually deployed to S3, but it seems like this is not needed here)

Reply:
So originally the way AppSync was created was that we supported reading the `graphql` file. So in creating the static function that read a local file, we decided to follow the same scheme as Lambda `Code.fromAsset(...)`. Im guessing Asset is a misleading word?

Reply:
A little bit, in my opinion. "Asset" is a domain concept from the CDK (see [here](https://docs.aws.amazon.com/cdk/latest/guide/assets.html)).

I personally would have called that method `fromFile()`.

But let's maybe leave this one for now. It's kind of a separate issue. So I won't block the PR on it.

Reply:
Hm yeah okay I can live with that. If it makes your life easier, I can also split this PR into two to lower the file count!